[GStreamer][WebRTC] Allow backend to keep track of buffered-amount

philn opened the pull request:
#### af95424cb23375b928bd7f5cc1cd68706c131a99
<pre>
[GStreamer][WebRTC] Allow backend to keep track of buffered-amount
<a href="https://bugs.webkit.org/show_bug.cgi?id=235879">https://bugs.webkit.org/show_bug.cgi?id=235879</a>
&lt;rdar://problem/88561869&gt;

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer data-channel handler now monitors the underlying buffered-amount property and notifies
the client when it decreases. The GstWebRTCDataChannel on-buffered-amount-low cannot be used because
the final decision on firing the bufferedAmountIsDecreasing event depends on the bufferedAmount
value managed by the client (RTCDataChannel).

* Source/WebCore/Modules/mediastream/RTCDataChannel.h:
* Source/WebCore/Modules/mediastream/RTCDataChannelRemoteSource.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
(WebCore::GStreamerDataChannelHandler::GStreamerDataChannelHandler):
(WebCore::GStreamerDataChannelHandler::setBufferedAmountLowThreshold):
(WebCore::GStreamerDataChannelHandler::bufferedAmountChanged):
(WebCore::GStreamerDataChannelHandler::onBufferedAmountLow): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.h:
* Source/WebCore/platform/mediastream/RTCDataChannelHandler.h:
(WebCore::RTCDataChannelHandler::setBufferedAmountLowThreshold):
* Source/WebCore/platform/mediastream/RTCDataChannelHandlerClient.h:
(WebCore::RTCDataChannelHandlerClient::bufferedAmount const):

Canonical link: <a href="https://commits.webkit.org/251990@main">https://commits.webkit.org/251990@main</a>
</pre>

youennf commented on Source/WebCore/platform/mediastream/RTCDataChannelHandler.h:
```diff
@@ -115,6 +115,7 @@ class RTCDataChannelHandler {
 
     virtual void setClient(RTCDataChannelHandlerClient&, ScriptExecutionContextIdentifier) = 0;
 
+    virtual void setBufferedAmountLowThreshold(size_t) = 0;
```
Maybe set it as { } instead of pure virtual, so that only GST has to override this one.

youennf commented on Source/WebCore/Modules/mediastream/RTCDataChannel.cpp:
```diff
@@ -248,10 +254,15 @@ void RTCDataChannel::didDetectError(Ref<RTCError>&& error)
             dispatchEvent(Event::create(eventNames().bufferedamountlowEvent, Event::CanBubble::No, Event::IsCancelable::No));
+#else
+        UNUSED_VARIABLE(amount);
+        dispatchEvent(Event::create(eventNames().bufferedamountlowEvent, Event::CanBubble::No, Event::IsCancelable::No));
```
It seems that what you are after is something like bufferedAmountIsBelowThreshold.
It might be clearer to introduce that one instead of reusing bufferedAmountIsDecreasing with a dummy amount.

I also wonder how you will be able to implement the spec, which is based on a bufferedAmount which is manipulated on the context's thread, and not the backend background thread.
For instance, if the backend hops to the main thread, they need to call queueTaskKeepingObjectAlive.
Between the time queueTaskKeepingObjectAlive is called and the time the event is fired, the web app might have called send multiple times, which will have updated m_bufferedAmount.
How will you cancel firing the event in that case?

philn replied:
You're right, this is not going to be spec-compliant. I really need to have GStreamer pass the `amount` back to us. I'll revert this part of the patch.

philn replied:
So I suspect GstWebRTC is not exactly spec-compliant about this. I think it can be fixed there. So then, whenever we receive a `onbufferedamountlow` event from GstWebRTC on the main thread, we could directly bubble this up to JS side.

The racy behaviour you describe would be avoided I think. The buffered-amount measurements being done correctly in GstWebRTC would guarantee that the `onbufferedamountlow` is always emitted only when needed...

philn replied:
Actually I need this part still, because in our case `amount` is 0... I doubt a patch will be accepted wiring up an `amount` value from GstWebRTC to the app because it's not part of the public spec API surface...

youennf replied:
I do not see how GstWebRTC can avoid the racy behavior, given we call queueTaskKeepingObjectAlive which will dispatch the event asynchronously.
If we want to avoid the raciness, we would need GstWebRTC to compute the buffered amount in the JS thread, and dispatch the event synchronously without the queueTaskKeepingObjectAlive.
Maybe all this can be moved to GstWebRTC specific code (GStreamerDataChannelHandler.cpp).

Also m_bufferedAmount would never be decreased in your implementation, which is observable from JS (RTCDataChannel as a bufferedAmount getter). Firing the event with a bad bufferedAmount does not seem great.

> I doubt a patch will be accepted wiring up an `amount` value from GstWebRTC to the app because it's not part of the public spec API surface.

It is: https://w3c.github.io/webrtc-pc/#dom-datachannel-bufferedamount

youennf replied:
If GstWebRTC has a way to get bufferedAmount, it could be monitor this way.
Or WebKit GstWebRTC class could try to generate the low memory signal if it knows when a message has been sent?

philn replied:
Yes, GstWebRTCDataChannel has a `buffered-amount` property which we can monitor. The `amount` I was referring to previously is the delta between new amount and previous amount. Is that exposed in the spec? I thought it wasn't.

I'll update the patch and try another approach for events dispatching...

youennf replied:
> The amount I was referring to previously is the delta between new amount and previous amount. Is that exposed in the spec?

That one is indeed not exposed.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
```diff
@@ -224,6 +223,38 @@ void GStreamerDataChannelHandler::readyStateChanged()
+    auto notifyClient = [&](uint64_t amount) {
+        GST_DEBUG("Buffered amount getting low on channel %p", m_channel.get());
+        if (!m_client) {
+            GST_DEBUG("No client yet on channel %p, queueing notification", m_channel.get());
```
I see no notification being queued here

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
```diff
@@ -224,6 +223,38 @@ void GStreamerDataChannelHandler::readyStateChanged()
+
+    GST_DEBUG("New buffered amount on channel %p: %" G_GUINT64_FORMAT " old: %" G_GUINT64_FORMAT, m_channel.get(), bufferedAmount, m_cachedBufferedAmount ? *m_cachedBufferedAmount : -1);
+
+    auto notifyClient = [&](uint64_t amount) {
```
You can probably inline this.

philn committed:
[GStreamer][WebRTC] Allow backend to keep track of buffered-amount
https://bugs.webkit.org/show_bug.cgi?id=235879
<rdar://problem/88561869>

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer data-channel handler now monitors the underlying buffered-amount property and notifies
the client when it decreases. The GstWebRTCDataChannel on-buffered-amount-low cannot be used because
the final decision on firing the bufferedAmountIsDecreasing event depends on the bufferedAmount
value managed by the client (RTCDataChannel).

* Source/WebCore/Modules/mediastream/RTCDataChannel.h:
* Source/WebCore/Modules/mediastream/RTCDataChannelRemoteSource.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
(WebCore::GStreamerDataChannelHandler::GStreamerDataChannelHandler):
(WebCore::GStreamerDataChannelHandler::setBufferedAmountLowThreshold):
(WebCore::GStreamerDataChannelHandler::bufferedAmountChanged):
(WebCore::GStreamerDataChannelHandler::onBufferedAmountLow): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.h:
* Source/WebCore/platform/mediastream/RTCDataChannelHandler.h:
(WebCore::RTCDataChannelHandler::setBufferedAmountLowThreshold):
* Source/WebCore/platform/mediastream/RTCDataChannelHandlerClient.h:
(WebCore::RTCDataChannelHandlerClient::bufferedAmount const):

Canonical link: https://commits.webkit.org/251990@main
